### DISCUSSION TITLE
Machine-ID & Role based Application Access

### DISCUSSION THREAD

**Discussion by bendoerr** [gravitational/teleport#21441] [Q&A]
I have a use-case where we would like to use machine-id/tbot to provide access to applications. However, at configuration time of the tbot.yml, we won't always know all of the applications which need to be accessed. I figured we could use the role-based access control to allow the bot to access applications based on tag, similar to how users function. I can specify a 'role' in the `destination` part of the config but that was not working. Machine-ID Application Access works as long as I specify as specific app in it's config.

Is this use case possible?

**Comment by strideynet** [comment]
Unfortunately, at the moment, applications must be configured ahead of time. This is because a field is encoded into the certificate in order to indicate to the proxy which application to direct the requests to.

This is definitely something I want to improve in the future.  One avenue I'm looking into has been exposing a gRPC API on `tbot` that consumers can call to retrieve a stream of certificates for a specific target. In order to validate this, I'd love to know if the intended consumer here is something you control the source of and would be able to modify to support fetching certificates dynamically from `tbot` ? How does this solution sound to you ?

**Reply by bendoerr** [reply]
Yes the consumer is something we control and would modify to support fetching certificates dynamically. We are currently doing this by generating a tbot.yml for each app we end up needing and running tbot with `--oneshot`.

**Reply by strideynet** [reply]
Thanks for that - I'll make sure we've got an issue raised with the specifics and link it here for you to track :)

**Reply by strideynet** [reply]
https://github.com/gravitational/teleport/issues/14507

This ticket covers the general work of refactoring `tbot` to support a more dynamic mode of operation.
